ir/memory.cpp

```diff
@@ -814,11 +810,25 @@ expr Pointer::block_val_refined(const Pointer &other) const {
                 val.ptr_byteoffset() == val2.ptr_byteoffset() &&
                 val.ptr().refined(val2.ptr());
   }
+  // allow null ptr <-> zero
+  // If ptr points to a null block with ofs, its integer address is compared.
+  // The last bits of offsets are compared for approximation.
+  // Relevant transformations:
+  //   store ty* null  <-> memset i8 0
+  //   store ty* undef  -> memset i8 undef
+  expr val_i8 = expr::mkIf(is_ptr,
+                           val.ptr().get_offset().zextOrTrunc(bits_byte),
+                           val.nonptr_value());
+  expr val2_i8 = expr::mkIf(is_ptr2,
+                            val2.ptr().get_offset().zextOrTrunc(bits_byte),
+                            val2.nonptr_value());
+  expr null_punning = (!is_ptr2 || val2.ptr().isNull(false)) &&
+                      (!is_ptr || val.ptr().isNull(false)) &&
+                      !val2.is_poison() && val_i8 == val2_i8;
```

Comment:
When this code is used, either is_ptr is true or is_ptr2 is true, but not both.
So I think we can make this constraint simpler, e.g.:
```
!val2.is_poison() &&
  mkif(is_ptr,
       val.ptr().isNull(false) && val.ptr().get_offset().zextOrTrunc(bits_byte) == val2.nonptr_value(),
       val2.ptr().isNull(false) && val.nonptr_value() == val2.ptr().get_offset().zextOrTrunc(bits_byte))
```

Though this feels a bit strange to me. Why are we comparing an 8-bit value with a 64-bit offset truncated to 8 bits? Seems wrong. And this is not detecting undef at all, right?
This will allow something like (undesired):
```
%p = gep i8, null, 42
store %p, %q
=>
store 42, %q
store 42, %q+1
store 42, %q+2
store 42, %q+...
```

Reply:
Detecting undef seemed hard because the generated formula looked like this:
(select (store .. (store init_mem undef) undef) |idx|)
This cannot be constant folded because idx isn't constant.

Regarding the truncation, it was my suggested approximation. Otherwise we need to shift the corresponding bits, which seemed too expensive for resolving the single case from undef.

Reply:
To supplement: the added refinement here is nullptr -> int, and undefptr -> nullptr is already allowed, so undefptr -> nullptr -> int is now allowed.

Reply:
right, but we need to make sure we don't introduce false negatives.
Let's think a bit more in a better solution. I don't think the solution you're proposing is a good tradeoff. Better 1 false positive than introducing false-negatives IMHO.